src/main/java/org/sonarlint/intellij/inspection/SonarLintGlobalInspection.java

```diff
@@ -141,20 +144,61 @@ public void visitFile(PsiFile psiFile) {
   }
 
   ProblemDescriptor getProblemDescriptor(Issue issue, PsiFile psiFile, @NotNull InspectionManager manager, boolean isLocal) {
-    if (issue.getLine() != null) {
-      return manager.createProblemDescriptor(psiFile,
-        InspectionUtils.getLineRange(psiFile, issue.getLine()),
-        InspectionUtils.getProblemMessage(issue, isLocal),
-        problemHighlightType(issue, isLocal),
-        false);
+    Integer startLine = issue.getStartLine();
+    if (startLine != null) {
+      Integer startLineOffset = issue.getStartLineOffset();
+      if (startLineOffset != null) {
+        Integer endLine = issue.getEndLine();
+        Integer endLineOffset = issue.getEndLineOffset();
+        int ijStartLine = startLine - 1;
+        int ijEndLine = endLine - 1;
+        final Document document = PsiDocumentManager.getInstance(psiFile.getProject()).getDocument(psiFile);
+        PsiElement startElement;
+        PsiElement endElement;
+        if (document != null) {
+          int lineStartOffset = document.getLineStartOffset(ijStartLine);
+          final int startOffset = lineStartOffset + startLineOffset;
+          final int endOffset = (ijEndLine == ijStartLine) ? lineStartOffset + endLineOffset : document.getLineStartOffset(ijEndLine) + endLineOffset;
+          startElement = psiFile.getViewProvider().findElementAt(startOffset);
+          if (startElement == null || startElement.getTextRange().getStartOffset() != startOffset) {
+            return createProblemDescriptorPreciseTextRange(issue, psiFile, manager, isLocal, new TextRange(startOffset, endOffset));
+          } else {
+            endElement = psiFile.getViewProvider().findElementAt(endOffset - 1);
+            if (endElement == null ||  endElement.getTextRange().getEndOffset() != endOffset)  {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS134)